pywhy_graphs/algorithms/generic.py

```diff
@@ -333,3 +334,236 @@ def _single_shortest_path_early_stop(G, firstlevel, paths, cutoff, join, valid_p
+
+    Returns
+    -------
+    out : set
+        The parents of the provided node.
+    """
+
+    return nx.ancestors(G.sub_directed_graph(), node)
+
+
+def _directed_sub_graph_parents(G, node):
+    """Finds the parents of a node in the directed subgraph.
+
+    Parameters
+    ----------
+    G : Graph
+        The graph.
+    node : node label
+        The node for which we have to find the parents.
+
+    Returns
+    -------
+    out : set
+        The parents of the provided node.
+    """
+
+    return set(G.sub_directed_graph().predecessors(node))
+
+
+def _bidirected_sub_graph_neighbors(G, node):
+    """Finds the neighbors of a node in the bidirected subgraph.
+
+    Parameters
+    ----------
+    G : Graph
+        The graph.
+    node : node label
+        The node for which we have to find the neighbors.
+
+    Returns
+    -------
+    out : set
+        The parents of the provided node.
+    """
+    bidirected_parents = set()
+
+    if not isinstance(G, CPDAG):
+        bidirected_parents = set(G.sub_bidirected_graph().neighbors(node))
+
+    return bidirected_parents
+
+
+def _is_collider(G, prev_node, cur_node, next_node):
+    """Checks if the given node is a collider or not.
+
+    Parameters
+    ----------
+    G : graph
+        The graph.
+    prev_node : node
+        The previous node in the path.
+    cur_node : node
+        The node to be checked.
+    next_node:
+        The next node in the path.
+
+    Returns
+    -------
+    iscollider : bool
+        Bool is set true if the node is a collider, false otherwise.
+    """
+    parents = _directed_sub_graph_parents(G, cur_node)
+    parents = parents.union(_bidirected_sub_graph_neighbors(G, cur_node))
+
+    if prev_node in parents and next_node in parents:
+        return True
+
+    return False
+
+
+def _recursive_path(G, node_x, node_y, L, S, visited, check_ancestors, cur_node, prev_node):
+    """Recursively explores a graph to find a path.
+
+       Finds path that are compliant with the inducing path requirements.
+
+    Parameters
+    ----------
+    G : graph
+        The graph.
+    node_x : node
+        The source node.
+    node_y : node
+        The destination node
+    L : set
+        Set containing all the non-colliders.
+    S : set
+        Set containing all the colliders.
+    visited : set
+        Set containing all the nodes already visited.
+    check_ancestors : set
+        Set containing all the ancestors to be checked against.
+    cur_node : node
+        The current node.
+
+    Returns
+    -------
+    path : Tuple[bool, path]
+        A tuple containing a bool and a path which is empty if the bool is false.
+    """
+    path_exists = False
+    path = []
+    visited.add(cur_node)
+    children = G.neighbors(cur_node)
+
+    if cur_node is node_y:
+        return (True, [node_y])
+
+    for elem in children:
+        if elem in visited:
+            continue
+
+        else:
+            if (
```

Comment:
```suggestion
# if the current node is a collider, check that...
            if (
```

Add in-line comments to make it clear what this complex if/else sequence is doing